Add styling to distinguish bots from human users.

tommyip opened the pull request:
Wrap bot's name inside a bootstrap label with a font-awesome android icon.
Fixes #1107

It is ok to use the Andriod icon right? (any legal issues?) I think that is the only bot icon available in font-awesome. Also is the style ok?

Screenshot:
![bot-icons](https://cloud.githubusercontent.com/assets/18504232/21471948/dd741970-cabc-11e6-8340-a1c995720d73.png)

smarx commented:
*Automated message from Dropbox CLA bot*

@tommyip, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Awesome, thanks for working on this @tommyip!!!  I've been wanting to fix for this for a long time.  I had kinda been imagining putting the bot symbol after the name, rather than before; thoughts on that question are super welcome!

Using the android emoji I'd worry about being confusing, in that some users might think it means the message comes from an Android phone (for example).  So I would look for (or make) an alternative we can use for the purpose.  

@showell @brockwhittaker what do you think we should use as the label styling to indicate bots?  I worry the Bootstrap label may create too much emphasis in addition to having a bot symbol, but I'm somewhat flexible.

(We should also think try to propagate whatever change we make here to the admin pages, but best to settle the initial question of how to display these first).

tommyip commented:
Several other ideas:
![bot-icons1](https://cloud.githubusercontent.com/assets/18504232/21499190/a683401a-cc2b-11e6-98cd-1ce28b94f8b5.png)
* I changed the green to #94c849 which should match with the green used in other places.
* Still couldn't find an bot font icon, any ideas where I could find one?

timabbott commented:
What do you think about just displaying the a bot icon after the name (maybe not in bold, so it's clearly different from a person whose name ends with that icon), without the bubble?

kevinr commented:
I don't love the existing bot icon. It doesn't read very well at that size---it mostly just looks likes a blob to me.  Anything else we find still runs the risk of being confused with an emoji in the user or bot's display name.

I think my preference is for something like the option third from the left in [this comment](https://github.com/zulip/zulip/pull/2909#issuecomment-269316879).

In particular, I prefer the bubble around the flair, since it is hard to forge in Unicode and calls attention to the flair without changing the emphasis or appearance of the display name.

We can imagine analogous flair for admins or moderators, et al.

timabbott commented:
We definitely want a different bot icon, but what I was thinking without the flare is that we'd make it a different color (e.g. a grey) which would be clearly unforgeable.

arpith commented:
Re: different icons, would one of these work?
https://thenounproject.com/term/robot/294999/
https://thenounproject.com/term/robot-face/332575/
https://thenounproject.com/term/robot-head/771016/

(from https://thenounproject.com/search/collection/?q=robot)

kevinr commented:
I'm not sure any of those are great. I'll tell you what my process was, so you can verify it yourself.

I opened up each of those pages, and then I zoomed the page out all the way so the icon was about as tall as the sender name text in Zulip (1/4 inch, or about 1 cm).

Here's what they look like:

<img width="26" alt="bot-icon-1" src="https://cloud.githubusercontent.com/assets/52371/21532547/cce8306c-cd16-11e6-827b-dccf0cdd55cc.png">
<img width="30" alt="bot-icon-2" src="https://cloud.githubusercontent.com/assets/52371/21532550/d2dc866c-cd16-11e6-85cb-3d5d3306ea85.png">
<img width="33" alt="bot-icon-3" src="https://cloud.githubusercontent.com/assets/52371/21532552/d75e2e0c-cd16-11e6-84b0-74061eacc720.png">

At that scale, it was very hard to tell what the icons represent.  I think the "face" gets lost.  It's at about twice that size that they start being legible.

Text is less ambiguous.

timabbott commented:
Text is uglier, though.  I think with the right icon we can make this work.  Certainly the Android icon is recognizable at that scale; we can find a bot icon that is as well.

I think https://thenounproject.com/term/robot-face/332575/ is the most promising of those ideas, but I think is too rounded and thus doesn't convey the idea that well.

tommyip commented:
@timabbott icons are easy to recognize but @kevinr is right that it looks like a blob when in its normal size. Should I design a custom one?

timabbott commented:
Designing a custom one would be a totally reasonable approach!

tommyip commented:
I am no designer but I will try!

tommyip commented:
Just a quick prototype, is it still too clustered?
![bot64](https://cloud.githubusercontent.com/assets/18504232/21534841/d50c5c70-cd64-11e6-8d37-060db7b0b839.png)

timabbott commented:
Can you show a screenshot of just putting the but icon after the name, colored grey (with no special bubble and no text)?  I think that's the direction we want to go in with this...

tommyip commented:
@timabbott Like this?
![bot_grey](https://cloud.githubusercontent.com/assets/18504232/21574745/71439d10-cef0-11e6-99be-be3b84fe723e.png)

brockwhittaker commented:
@tommyip there's two quick issues:

1. The icon for the bot is too big as it goes to the top of the text and well below it as well.
2. It is/appears to be the same color of black as the author's name which means that it looks like in theory it could be a custom unicode character someone put in. I'd recommend making it a different color because at least that will then signal that it isn't standard unicode. Perhaps to a light grey like `#aaa`?

brockwhittaker commented:
Also, one other thing @tommyip, I would change your function into an IIFE so that it is clear that the contents of is_bot will be what they are regardless of whether or not the underlying objects it is retrieved from are changed. It also helps free up memory by not requiring a reference to objects that may otherwise be killed.

I would change it to:

```javascript
message_container.is_bot = (function is_bot() {
    var email = message_container.msg.sender_email;
    return people.get_by_email(email).is_bot;
}());
```

umairwaheed commented:
@tommyip, still working on this?

tommyip commented:
@umairwaheed yes I am.

tommyip commented:
@brockwhittaker @timabbott PR updated:
* Changed `is_bot()` to IIFE
* Simplified bot icon and limit the size to keep inline with other characters at the baseline height.
* Change color to #aaaaaa

![bot_icon](https://cloud.githubusercontent.com/assets/18504232/21855401/56252772-d816-11e6-9805-5bc390fe671c.png)

timabbott commented:
Nice, I think this achieves all of our goals visually -- it's clear what a bot is, the icon is pretty clear, and it's subtle enough to not draw attention when a user isn't trying to figure out what's up with the user.

timabbott commented on static/templates/single_message.handlebars:
```diff
@@ -15,6 +15,9 @@
               {{msg/sender_full_name}}
+              {{#if is_bot}}
+                <img class="bot-icon" src="/static/images/bot.png" />
+              {{/if}}
```
In theory we should have similar logic in the next case for handling bot messages starting with `/me`.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -218,6 +218,11 @@ MessageListView.prototype = {
+            message_container.is_bot = (function is_bot() {
+                var email = message_container.msg.sender_email;
+                return people.get_by_email(email).is_bot;
+            }());
```
I would probably move this up to be next to the `sender` logic; and also the variable should be called `sender_is_bot`, since it's the sender, not the message, that's a bot.

tommyip commented:
Fixed both issues!

tommyip commented:
@timabbott Ready for final review.

timabbott commented:
@tommyip awesome, thanks for your work on this, and sorry for the delayed review!  I think the new icon is pretty good, but I'm still struggling with whether this icon is too Darth Vadery; I'm going to ping some of our design folks on whether they have ideas for an icon that still clearly suggests "bot", but doesn't have that problem.  

So I've been holding off on merging this until after the 1.5 release so we can have lots of time to get feedback and play with the icon before this goes into a major release.  Remind me if I forget!

timabbott commented:
Merged as abf522adfb3ceabba5e5af4b0908327feda0928f, thanks @tommyip!  I'm really excited about this one :)
